### PR TITLE
Add flutter analyze because it is much faster

### LINKED ISSUE #46: How can we speedup sonar-flutter?  🚗💨
I'm currently investigating how we can make the sonar-flutter plugin faster by a large margin.

I read #30 by @EnPython and also found out about the PR #29 by @kuhnroyal, which tries to speed up sonar-flutter by using an existing analyzer report file.

After trying out the PR by @kuhnroyal (btw: awesome work! 🎉 ) and running some sonar checks locally, I found out that this indeed speeds up the Sonar check by a large margin.

As a downside, this requires having some sort of `analysis.txt` file generated by some command in the first place. That is why I tried to change the sonar-flutter code to run `flutter analyze` instead of running `dartanalyzer file1 file2 file3 ...`, and the results for our Flutter project was shockingly fast: it took ~23s instead of ~6m, which was a speedup of ~15x.

Of course, the issues found by `flutter analyze` and `dartanalyzer` can be slightly different - and maybe sonar-flutter projects are used by both Flutter and (pure) Dart developers.

That's why I want to know from you: which implementation would you prefer?

1. Change the default from `dartanalyzer` to `flutter analyze`: This would make analysis faster for everyone (yay), but could potentially break for pure-Dart devs
2. Create a setting enum LEGACY/FLUTTER: People could choose the faster Flutter analysis from Sonar settings - but they have to explicitly opt-in (so some people will never know about this option being available)
3. Add an automated check that checks `pubspec.yaml` if Flutter is available and runs `flutter analyze` for Flutter projects and `dartanalyzer` for all others: That way Flutter devs would benefit instantly, and Dart devs would still get the old analysis
4. Some idea of you :)

I'd gladly contribute a PR with this - but it's good to know which direction you would prefer first :)

### PR DESCRIPTION
**Pull request by mreichelt**
Fixes #46

### CODE REVIEW

**Review by kuhnroyal — APPROVED** [review]

File: dart-lang/src/main/java/fr/insideapp/sonarqube/dart/lang/issues/dartanalyzer/AnalyzerMode.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package fr.insideapp.sonarqube.dart.lang.issues.dartanalyzer;
+
+public enum AnalyzerMode {
+    flutter,
+    legacy;
```
[the remark this answers is not in the document]
Comment (gwennguihal) [reply dart-lang/src/main/java/fr/insideapp/sonarqube/dart/lang/issues/dartanalyzer/AnalyzerMode.java:26]: I just saw your last commit, thanks ;)

[the remark this answers is not in the document]
Comment (mreichelt) [reply dart-lang/src/main/java/fr/insideapp/sonarqube/dart/lang/issues/dartanalyzer/AnalyzerMode.java:26]: Yeah, I added the `dart` one as well based on the suggestion by @kuhnroyal :)

To clarify:

- `legacy` option uses the `dartanalyzer` command (the 'old' way)
- `flutter` option uses `flutter analyze` (much faster, also works for pure Dart projects when flutter is installed and available in PATH
- `dart` option uses `dart analyze` (much faster than legacy as well, but doesn't require Flutter SDK)

File: dart-lang/src/main/java/fr/insideapp/sonarqube/dart/lang/issues/dartanalyzer/DartAnalyzerSensor.java [hunk 2]
```diff
@@ -46,243 +46,293 @@
...
+        if (analysisOptionsFile.exists()) {
+            analysisOptionsFile
+                    .renameTo(new File(sensorContext.fileSystem().baseDir(), ANALYSIS_OPTIONS_FILENAME + ".sonar"));
+            LOGGER.info("Backup of original analysis_options.yaml file to {}", ANALYSIS_OPTIONS_FILENAME + ".sonar");
+        }
+    }
+
+    private void createAnalysisOptionsFile(SensorContext sensorContext) throws IOException {
+        File analysisOptionsFile = new File(sensorContext.fileSystem().baseDir(), ANALYSIS_OPTIONS_FILENAME);
+        URL inputUrl = getClass().getResource(ANALYSIS_OPTIONS_FILE);
```
Comment (kieraCurtis) [inline dart-lang/src/main/java/fr/insideapp/sonarqube/dart/lang/issues/dartanalyzer/DartAnalyzerSensor.java:338]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [UI: Usage of GetResource may be unsafe if class is extended (UI_INHERITANCE_UNSAFE_GETRESOURCE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#ui-usage-of-getresource-may-be-unsafe-if-class-is-extended-ui-inheritance-unsafe-getresource).
Calling this.getClass().getResource(...) could give results other than expected if this class is extended by a class in another package.

Comment (matpag) [reply dart-lang/src/main/java/fr/insideapp/sonarqube/dart/lang/issues/dartanalyzer/DartAnalyzerSensor.java:338]: @kieraCurtis Would be `getClass.getClassLoader().getResource(...)` a better option?

Comment (jrehwaldt) [reply dart-lang/src/main/java/fr/insideapp/sonarqube/dart/lang/issues/dartanalyzer/DartAnalyzerSensor.java:338]: What about `DartAnalyzerSensor.class.getResource(...)`?

Comment (matpag) [reply dart-lang/src/main/java/fr/insideapp/sonarqube/dart/lang/issues/dartanalyzer/DartAnalyzerSensor.java:338]: @jrehwaldt Good point, probably it's the right choice here

Comment (matpag) [reply dart-lang/src/main/java/fr/insideapp/sonarqube/dart/lang/issues/dartanalyzer/DartAnalyzerSensor.java:338]:
```suggestion
        URL inputUrl = DartAnalyzerSensor.class.getResource(ANALYSIS_OPTIONS_FILE);
```
@mreichelt Maybe you can give it a try when you have time

Comment (mreichelt) [reply dart-lang/src/main/java/fr/insideapp/sonarqube/dart/lang/issues/dartanalyzer/DartAnalyzerSensor.java]: It's good to have multiple eyes on this, thanks @kieraCurtis, @matpag and @jrehwaldt. I added the change. 👍

To be fair, I didn't add that line of code in the first place - it was all there all along. You can hide the whitespace changes to see what I actually changed like so:
https://github.com/insideapp-oss/sonar-flutter/pull/47/files?diff=unified&w=1

@zippy1978 and @victorgilc: it seems you are pretty busy with other things - I hope that you are well and that you didn't give up on this project.

**Comment by kuhnroyal** [comment]
So, after testing my `flutter analyze --write=analysis_results.txt` output today, I found that it is different from `dartanalyzer > analysis_results.txt` and also different from `dart analyze > analysis_results.txt`, well an also from `flutter analyze > analysis_results.txt`.

So there are 4 (I think) different output formats, 3 of wich can be parsed atm.
The `flutter analyze --write` is currently not supported, only the redirected outputs work.

File: dart-lang/src/main/java/fr/insideapp/sonarqube/dart/lang/issues/dartanalyzer/DartAnalyzerReportParser.java [hunk 3]
```diff
@@ -31,7 +31,7 @@
         List<DartAnalyzerReportIssue> issues = new ArrayList<>();
 
         String[] lines = input.split(System.getProperty("line.separator"));
-        Pattern pattern = Pattern.compile("(error|hint|lint)(.*)(-|•)(.*)(-|•)(.*):(.*):(.*)(-|•)(.*)");
+        Pattern pattern = Pattern.compile("(hint|lint|info|warning|error)(.*)(-|•)(.*)(-|•)(.*):(.*):(.*)(-|•)(.*)");
```
Comment (kuhnroyal) [inline dart-lang/src/main/java/fr/insideapp/sonarqube/dart/lang/issues/dartanalyzer/DartAnalyzerReportParser.java:34]: I think we need different patterns here, per analyzer command option :(

**Comment by kuhnroyal** [comment]
Ok some more info for this:
* there is now a `--format=machine` flag for `dartanalyzer` and `dart analyze`, both produce the same output
```
INFO|HINT|UNUSED_ELEMENT|/Users/.../example/lib/src/example_form.dart|8|7|4|The declaration '_foo' isn't referenced.
INFO|LINT|PREFER_SINGLE_QUOTES|/Users/.../example/lib/src/example_form.dart|8|14|2|Only use double quotes for strings containing single quotes.
```
* this is good, we should probably support this as the new default and also update the readme 
* `flutter analyze` does not support this flag yet, but there is https://github.com/flutter/flutter/projects/106 with several issue about adding the support

Sadly I am primarily using `flutter analyze`

**Comment by wreckitq** [comment]
it's fixed this issue https://github.com/insideapp-oss/sonar-flutter/issues/38

**Comment by wreckitq** [comment]
thanks so much @mreichelt

**Comment by zippy1978** [comment]
Thanks a lot @mreichelt !
I just merged it to develop

**Comment by wreckitq** [comment]
Finally 🥳

**Comment by bogdannedelcu** [comment]
I am getting the error Could not startup process 'flutter analyze' while flutter is in my path, and works if i write in the command line "flutter analyze".

using macos, 0.3.2 version of the plugin. Thank you!


$ sonar-scanner   -Dsonar.projectKey=Flow   -Dsonar.sources=lib   -Dsonar.host.url=http://localhost:9000   -Dsonar.login=6456683daa4b24dfdf1dfa990814088a56eca343 -Dsonar.sourceEncoding=UTF-8 -Dsonar.dart.analysis.useExistingOptions=false 
INFO: Scanner configuration file: /Users/andreea/sonar-scanner/conf/sonar-scanner.properties
INFO: Project root configuration file: NONE
INFO: SonarScanner 4.6.1.2450
INFO: Java 11.0.3 AdoptOpenJDK (64-bit)
INFO: Mac OS X 10.16 x86_64
INFO: User cache: /Users/andreea/.sonar/cache
INFO: Scanner configuration file: /Users/andreea/sonar-scanner/conf/sonar-scanner.properties
INFO: Project root configuration file: NONE
INFO: Analyzing on SonarQube server 8.8.0
INFO: Default locale: "ro_RO", source code encoding: "UTF-8"
INFO: Load global settings
INFO: Load global settings (done) | time=66ms
INFO: Server id: BF41A1F2-AXkt3uU8s7ScmhUKsoCU
INFO: User cache: /Users/andreea/.sonar/cache
INFO: Load/download plugins
INFO: Load plugins index
INFO: Load plugins index (done) | time=28ms
INFO: Load/download plugins (done) | time=74ms
INFO: Process project properties
INFO: Process project properties (done) | time=7ms
INFO: Execute project builders
INFO: Execute project builders (done) | time=1ms
INFO: Project key: Flow
INFO: Base dir: /Users/andreea/git_projects/Flow-Flutter
INFO: Working dir: /Users/andreea/git_projects/Flow-Flutter/.scannerwork
INFO: Load project settings for component key: 'Flow'
INFO: Load project settings for component key: 'Flow' (done) | time=9ms
INFO: Load quality profiles
INFO: Load quality profiles (done) | time=32ms
INFO: Load active rules
INFO: Load active rules (done) | time=1099ms
INFO: Indexing files...
INFO: Project configuration:
INFO: 184 files indexed
INFO: 0 files ignored because of scm ignore settings
INFO: Quality profile for dart: dartanalyzer
INFO: ------------- Run sensors on module Flow
INFO: Load metrics repository
INFO: Load metrics repository (done) | time=32ms
INFO: Sensor CSS Rules [cssfamily]
INFO: No CSS, PHP, HTML or VueJS files are found in the project. CSS analysis is skipped.
INFO: Sensor CSS Rules [cssfamily] (done) | time=0ms
INFO: Sensor JaCoCo XML Report Importer [jacoco]
INFO: 'sonar.coverage.jacoco.xmlReportPaths' is not defined. Using default locations: target/site/jacoco/jacoco.xml,target/site/jacoco-it/jacoco.xml,build/reports/jacoco/test/jacocoTestReport.xml
INFO: No report imported, no coverage information will be imported by JaCoCo XML Report Importer
INFO: Sensor JaCoCo XML Report Importer [jacoco] (done) | time=3ms
INFO: Sensor C# Project Type Information [csharp]
INFO: Sensor C# Project Type Information [csharp] (done) | time=1ms
INFO: Sensor C# Properties [csharp]
INFO: Sensor C# Properties [csharp] (done) | time=0ms
INFO: Sensor Dart sensor [flutter]
INFO: Load project repositories
INFO: Load project repositories
INFO: Load project repositories
INFO: Load project repositories (done) | time=52ms
INFO: Load project repositories (done) | time=53ms
INFO: Load project repositories (done) | time=30ms
INFO: Sensor Dart sensor [flutter] (done) | time=4541ms
INFO: Sensor dartanalyzer sensor [flutter]
INFO: Chosen analyzer mode: flutter
INFO: Running 'flutter analyze'...
ERROR: org.buildobjects.process.StartupException: Could not startup process 'flutter analyze'.
java.io.IOException: org.buildobjects.process.StartupException: Could not startup process 'flutter analyze'.
	at fr.insideapp.sonarqube.dart.lang.issues.dartanalyzer.DartAnalyzerSensor.getIssuesFromAnalyzer(DartAnalyzerSensor.java:137)
	at fr.insideapp.sonarqube.dart.lang.issues.dartanalyzer.DartAnalyzerSensor.recordIssuesFromAnalyzer(DartAnalyzerSensor.java:117)
	at fr.insideapp.sonarqube.dart.lang.issues.dartanalyzer.DartAnalyzerSensor.execute(DartAnalyzerSensor.java:103)
	at org.sonar.scanner.sensor.AbstractSensorWrapper.analyse(AbstractSensorWrapper.java:48)
	at org.sonar.scanner.sensor.ModuleSensorsExecutor.execute(ModuleSensorsExecutor.java:85)
	at org.sonar.scanner.sensor.ModuleSensorsExecutor.lambda$execute$1(ModuleSensorsExecutor.java:59)
	at org.sonar.scanner.sensor.ModuleSensorsExecutor.withModuleStrategy(ModuleSensorsExecutor.java:77)
	at org.sonar.scanner.sensor.ModuleSensorsExecutor.execute(ModuleSensorsExecutor.java:59)
	at org.sonar.scanner.scan.ModuleScanContainer.doAfterStart(ModuleScanContainer.java:82)
	at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:137)
	at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:123)
	at org.sonar.scanner.scan.ProjectScanContainer.scan(ProjectScanContainer.java:392)
	at org.sonar.scanner.scan.ProjectScanContainer.scanRecursively(ProjectScanContainer.java:388)
	at org.sonar.scanner.scan.ProjectScanContainer.doAfterStart(ProjectScanContainer.java:357)
	at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:137)
	at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:123)
	at org.sonar.scanner.bootstrap.GlobalContainer.doAfterStart(GlobalContainer.java:149)
	at org.sonar.core.platform.ComponentContainer.startComponents(ComponentContainer.java:137)
	at org.sonar.core.platform.ComponentContainer.execute(ComponentContainer.java:123)
	at org.sonar.batch.bootstrapper.Batch.doExecute(Batch.java:72)
	at org.sonar.batch.bootstrapper.Batch.execute(Batch.java:66)
	at org.sonarsource.scanner.api.internal.batch.BatchIsolatedLauncher.execute(BatchIsolatedLauncher.java:46)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source)
	at java.base/java.lang.reflect.Method.invoke(Unknown Source)
	at org.sonarsource.scanner.api.internal.IsolatedLauncherProxy.invoke(IsolatedLauncherProxy.java:60)
	at com.sun.proxy.$Proxy0.execute(Unknown Source)
	at org.sonarsource.scanner.api.EmbeddedScanner.doExecute(EmbeddedScanner.java:189)
	at org.sonarsource.scanner.api.EmbeddedScanner.execute(EmbeddedScanner.java:138)
	at org.sonarsource.scanner.cli.Main.execute(Main.java:112)
	at org.sonarsource.scanner.cli.Main.execute(Main.java:75)
	at org.sonarsource.scanner.cli.Main.main(Main.java:61)
Caused by: org.buildobjects.process.StartupException: Could not startup process 'flutter analyze'.
	at org.buildobjects.process.Proc.<init>(Proc.java:91)
	at org.buildobjects.process.ProcBuilder.run(ProcBuilder.java:205)
	at fr.insideapp.sonarqube.dart.lang.issues.dartanalyzer.DartAnalyzerSensor.getIssuesFromAnalyzer(DartAnalyzerSensor.java:130)
	... 32 common frames omitted
Caused by: java.io.IOException: Cannot run program "flutter": error=2, No such file or directory
	at java.base/java.lang.ProcessBuilder.start(Unknown Source)
	at java.base/java.lang.ProcessBuilder.start(Unknown Source)
	at java.base/java.lang.Runtime.exec(Unknown Source)
	at org.buildobjects.process.Proc.<init>(Proc.java:77)
	... 34 common frames omitted
Caused by: java.io.IOException: error=2, No such file or directory
	at java.base/java.lang.ProcessImpl.forkAndExec(Native Method)
	at java.base/java.lang.ProcessImpl.<init>(Unknown Source)
	at java.base/java.lang.ProcessImpl.start(Unknown Source)
	... 38 common frames omitted

INFO: Sensor dartanalyzer sensor [flutter] (done) | time=14ms
INFO: Sensor Flutter unit tests [flutter]
WARN: Test report file /Users/andreea/git_projects/Flow-Flutter/tests.output not found
INFO: Sensor Flutter unit tests [flutter] (done) | time=1ms
INFO: Sensor Flutter Coverage [flutter]
WARN: No coverage information will be saved because LCOV file cannot be found.
WARN: Provided LCOV file path: coverage/lcov.info. Seek file with path: /Users/andreea/git_projects/Flow-Flutter/coverage/lcov.info
INFO: Analysing [null]
INFO: Sensor Flutter Coverage [flutter] (done) | time=3ms
INFO: Sensor JavaXmlSensor [java]
INFO: Sensor JavaXmlSensor [java] (done) | time=1ms
INFO: Sensor HTML [web]
INFO: Sensor HTML [web] (done) | time=6ms
INFO: Sensor VB.NET Project Type Information [vbnet]
INFO: Sensor VB.NET Project Type Information [vbnet] (done) | time=1ms
INFO: Sensor VB.NET Properties [vbnet]
INFO: Sensor VB.NET Properties [vbnet] (done) | time=1ms
INFO: ------------- Run sensors on project
INFO: Sensor Zero Coverage Sensor
INFO: Sensor Zero Coverage Sensor (done) | time=3ms
INFO: CPD Executor 7 files had no CPD blocks
INFO: CPD Executor Calculating CPD for 177 files
INFO: CPD Executor CPD calculation finished (done) | time=94ms
INFO: Analysis report generated in 169ms, dir size=1 MB
INFO: Analysis report compressed in 513ms, zip size=439 KB
INFO: Analysis report uploaded in 28ms
INFO: ANALYSIS SUCCESSFUL, you can browse http://localhost:9000/dashboard?id=Flow
INFO: Note that you will be able to access the updated dashboard once the server has processed the submitted analysis report
INFO: More about the report processing at http://localhost:9000/api/ce/task?id=AXkufET-VSd950CLEWO1
INFO: Analysis total time: 8.685 s
INFO: ------------------------------------------------------------------------
INFO: EXECUTION SUCCESS
INFO: ------------------------------------------------------------------------
INFO: Total time: 9.457s
INFO: Final Memory: 6M/37M
INFO: ------------------------------------------------------------------------
`

**Comment by bogdannedelcu** [comment]
my bad, I had the flutter/bin in my PATH starting with ~
After replacing with full name of the folder and restarted it all worked fine, both flutter analyze and dart analyze.

**Comment by mreichelt** [comment]
@zippy1978 Thank you so much for merging this, and for providing this awesome project! In my first test with version 0.3.2, the first Sonar run took 1m 7s instead of the previous 12m run! 🎉 

This is super awesome.

**Pull request merged**